CRM/Membershipextrasimporterapi/EntityImporter/Membership.php

```diff
@@ -0,0 +1,169 @@
+    $startDate = $this->formatRowDate('membership_start_date', 'Start Date', TRUE);
+    $endDate = $this->formatRowDate('membership_end_date', 'End Date', TRUE);
+
+    return [
+      1 => [$this->contactId, 'Integer'],
+      2 => [$membershipTypeId, 'Integer'],
+      3 => [$joinDate, 'Date'],
+      4 => [$startDate, 'Date'],
+      5 => [$endDate, 'Date'],
+      6 => [$membershipStatusId, 'Integer'],
+      7 => [1, 'Integer'],
+      8 => [$this->recurContributionId, 'Integer'],
+      9 => [$isOverriddenStatus, 'Integer'],
+      10 => [$statusOverrideEndDate, 'Date'],
+    ];
+  }
+
+  private function getMembershipTypeId() {
+    if (empty($this->rowData['membership_type'])) {
+      throw new CRM_Membershipextrasimporterapi_Exception_InvalidMembershipFieldException('Membership type is required field', 100);
+    }
+
+    if (!isset($this->cachedValues['membership_types'])) {
+      $sqlQuery = "SELECT id, name FROM civicrm_membership_type";
+      $result = CRM_Core_DAO::executeQuery($sqlQuery);
+      while ($result->fetch()) {
+        $this->cachedValues['membership_types'][$result->name] = $result->id;
+      }
+    }
+
+    if (!empty($this->cachedValues['membership_types'][$this->rowData['membership_type']])) {
+      return $this->cachedValues['membership_types'][$this->rowData['membership_type']];
+    }
+
+    throw new CRM_Membershipextrasimporterapi_Exception_InvalidMembershipFieldException('Invalid membership type', 200);
+  }
+
+  private function getMembershipStatusId($membershipTypeId) {
+    if (empty($this->rowData['membership_status'])) {
+      $statusId = CRM_Member_BAO_MembershipStatus::getMembershipStatusByDate(
+        $this->rowData['membership_start_date'],
+        $this->rowData['membership_end_date'],
+        $this->rowData['membership_join_date'],
+        'today',
+        TRUE,
+        $membershipTypeId
+      );
+
+      return $statusId['id'];
+    }
+
+    if (!isset($this->cachedValues['membership_statuses'])) {
+      $sqlQuery = "SELECT id, name FROM civicrm_membership_status";
+      $result = CRM_Core_DAO::executeQuery($sqlQuery);
+      while ($result->fetch()) {
+        $this->cachedValues['membership_statuses'][$result->name] = $result->id;
+      }
+    }
+
+    if (!empty($this->cachedValues['membership_statuses'][$this->rowData['membership_status']])) {
+      return $this->cachedValues['membership_statuses'][$this->rowData['membership_status']];
+    }
+
+    throw new CRM_Membershipextrasimporterapi_Exception_InvalidMembershipFieldException('Invalid membership status', 300);
+  }
+
+  private function getStatusOverrideEndDate() {
+    if (!empty($this->rowData['membership_status_override_end_date'])) {
+      return $this->formatRowDate('membership_status_override_end_date', 'Membership status override end date');
+    }
+
+    return NULL;
+  }
+
+  private function isOverriddenStatus($statusOverrideEndDate) {
+    if (!empty($statusOverrideEndDate)) {
+      return 2;
```

Comment:
Will overriden status always be 0, 1, 2? if yes, Should these number 0, 1, 2 in this function be constant variables with meaning of the number?

Reply:
so in PHP class variables are public and that cannot be changed, also this class is about importing so it does not make sense to have these defined here, Ideally, it should be defined as an enumeration (or since PHP does not support enumerations in a class that only contain them), but CiviCRM core already has a class that define these variables (called CRM_Member_StatusOverrideTypes), so I will just use it.